master/internal/rm/kubernetesrm/job.go

```diff
@@ -143,284 +140,363 @@ func newPod(
+				j.firstPodExit = &podExitInfo{
+					code: 1025,
+					msg:  "job exited with a failure but we don't have pod-level detail",
+				}
+			}
+			j.container.State = cproto.Terminated
+			j.informTaskResourcesStopped(sproto.ResourcesStopped{
+				Failure: &sproto.ResourcesRestoreError{
+					FailureType: sproto.ResourcesFailed,
+					ErrMsg:      j.firstPodExit.msg,
+					ExitCode:    (*sproto.ExitCode)(&j.firstPodExit.code),
+				},
+			})
+			return cproto.Terminated, nil
+		}
+	}
 
-	containerState, err := p.getPodState(p.pod, p.containerNames)
+	return j.container.State, nil
+}
+
+func (j *job) jobDeleted() {
+	exitCode := 1026
+	j.container.State = cproto.Terminated
+	j.informTaskResourcesStopped(sproto.ResourcesStopped{
+		Failure: &sproto.ResourcesRestoreError{
+			FailureType: sproto.ResourcesFailed,
+			ErrMsg:      "job was suddenly deleted but not by us (setting exit code 1026)",
+			ExitCode:    (*sproto.ExitCode)(&exitCode),
+		},
+	})
+}
+
+func (j *job) podStatusUpdate(updatedPod k8sV1.Pod) error {
+	j.mu.Lock()
+	defer j.mu.Unlock()
+
+	// TODO(TICKET): This is how the code was before but it doesn't make sense at all.
+	// spec.NodeName is a request to schedule on a specific node, not the node each pod is on.
+	// Needs a ticket.
+	j.podNodeNames[updatedPod.Name] = updatedPod.Spec.NodeName
+
+	podState, err := j.getPodState(updatedPod, j.containerNames)
 	if err != nil {
-		return p.container.State, err
+		return err
 	}
 
-	if containerState == p.container.State {
-		return p.container.State, nil
+	if podState == cproto.Running {
+		err := startPodLogStreamer(j.podInterface, updatedPod.Name, func(log []byte) {
+			j.receiveContainerLog(sproto.ContainerLog{
+				Timestamp: time.Now().UTC(),
+				RunMessage: &aproto.RunMessage{
+					Value:   string(log),
+					StdType: stdcopy.Stdout,
+				},
+			})
+		})
+		if err != nil {
+			return fmt.Errorf("starting pod logs streamer for %s: %w", updatedPod.Name, err)
+		}
+	}
+
+	if podState.Before(j.container.State) || podState == j.container.State {
+		return nil
 	}
 
-	switch containerState {
+	switch podState {
 	case cproto.Assigned:
 		// Don't need to do anything.
 
 	case cproto.Starting:
 		// Kubernetes does not have an explicit state for pulling container images.
 		// We insert it here because our  current implementation of the trial actor requires it.
-		p.syslog.Infof(
-			"transitioning pod state from %s to %s", p.container.State, cproto.Pulling)
-		p.container = p.container.Transition(cproto.Pulling)
-		p.informTaskResourcesState()
+		j.syslog.Infof("saw pod state go from %s to %s", j.container.State, cproto.Pulling)
+		j.container.State = cproto.Pulling
+		j.informTaskResourcesState()
 
-		p.syslog.Infof("transitioning pod state from %s to %s", p.container.State, containerState)
-		p.container = p.container.Transition(cproto.Starting)
-		p.informTaskResourcesState()
+		j.syslog.Infof("saw pod state go from %s to %s", j.container.State, cproto.Starting)
```

Comment:
unneeded change?

Reply:
i do need this now, let's make a note to chat about it. no change in this file is unneeded.